doc/developer/design/20231127_decoupled_isolated_coordinator.md

```diff
@@ -0,0 +1,487 @@
+
+- `WriteTs(timeline) -> Timestamp`: allocates and returns monotonically
+  increasing write timestamps.
+
+  This timestamp will be strictly greater than all prior values of `ReadTs()`
+  and `WriteTs`.
+
+- `ApplyWrite(timeline, timestamp)`: marks a write at time `timestamp` as
+  completed.
+
+  All subsequent values of `ReadTs()` will be greater or equal to the given
+  write `timestamp`.
+
+### CATALOG
+
+CATALOG can also be considered a shared service whose methods can be called
+from any component that needs it. And multiple actors can append data to it.
+
+Changes to the catalog are versioned in the `"catalog"` timeline, which is
+tracked separately from other timelines in TIMESTAMP ORACLE. Any successful
+append of updates to the catalog implicitly does an `ApplyWrite("catalog",
+new_upper)` on TIMESTAMP ORACLE.
+
+- `CompareAndAppend(expected_upper, new_upper, updates) -> Result<(),
+  AppendError>`: appends the given updates to the catalog _iff_ the expected
+  upper matches its current upper.
+
+- `SubscribeAt(timestamp) -> Subscribe`: returns a snapshot of the catalog at
+  `timestamp`, followed by an ongoing stream of subsequent updates.
+
+- `SnapshotAt(timestamp)`: returns an in-memory snapshot of the catalog at
+  `timestamp`.
+
+  A snapshot corresponds to our current in-memory representation of the catalog
+  that the Coordinator holds. This can be (and will be, in practice!)
+  implemented using `SubscribeAt` to incrementally keep an in-memory
+  representation of the catalog up to date with changes. However, it is
+  logically a different operation.
+
+### The controllers, aka. STORAGE and COMPUTE
+
+The controllers remain largely as they have been previously designed, but we
+introduce a distinction between deterministic controller command and
+ephemeral/query-like controller commands. See [Controller
+Commands](#controller-commands) above.
+
+The controllers can be seen as consuming a stream of deterministic controller
+commands, in the newly introduced `"catalog"` timeline, and
+`AdvanceCatalogFrontier` is downgrading its frontier.
+
+Ephemeral/query-like commands gain a new `catalog_as_of` parameter that tells
+the controller that it can only process them once its input frontier has
+sufficiently advanced.
+
+These changes make it so that the controller knows up to which CATALOG
+timestamp it has received commands and to ensure that peeks (or other
+query-like operations) are only served once we know that controller state is
+_at least_ up to date with the catalog state as of which the peek was
+processed. Note that we say the Controller has to be _at least_ up to date with
+the `catalog_as_of`, it is okay for its catalog frontier to be beyond that.
+This has the effect that in-flight peeks can end up in a situation where a
+Controller doesn't have the collection that we are trying to query, and we
+therefore have to report back an error. This is in line with current
+Materialize behavior.
```

Comment:
I have trouble reasoning about the correctness of this. If the controller were required to use the catalog state at _exactly_ `as_of`, that would make sense to me. As it stands, I think the correctness relies on some fact like collection IDs are immutable, and so the only possible three states for an ID are "collection as yet unknown", "collection exists with known definition", and "collection dropped", and so waiting for the catalog `as_of` to catch up eliminates the first state from consideration, and we're fine with either of the other states.

But what if collection definitions can change over time? I'm thinking about Parker's forthcoming work on `ALTER TABLE`? I'm not sure how this presents to the formal model of Materialize. Does `ALTER TABLE ... ADD COLUMN` in-place mutate the definition of an existing collection? Or does it mint a new collection with a new ID? (Certainly the current plans are to use in-place mutate the existing definition of the global ID, but we could say that a collection is logically identified by `(global_id, revision)` and the revision is bumped after an `ALTER TABLE`.)

Anyway, the point here is that a peek for `SELECT * FROM t` might violate serializability in a post-`ALTER TABLE` world if the only constraint on when the peek executes is that the peek's `catalog_as_of` is less than or equal to the controller's catalog. The peek also needs to be executed _before_ the controller learns of any future alterations to `t`.

That's all braindump-y. Let me know if anything is unclear and I can take another stab at explaining my thinking.

Reply:
> A peek cannot be allowed to hold up progress, it would require putting in place holds or locks, which seems antithetical to the idea of isolated peeks.

Oh this seems like a good thing to shake out! They already do establish a read hold on the data collection. And perhaps establishing a read hold on the catalog collection is not so crazy? Idk! It depends what goes wrong if you keep a hold on the catalog collection. Holding up a drop dataflow command until all outstanding peeks are drained seems potentially fine. Refusing to answer new peeks because an `ALTER TABLE` is stalled out on an ancient peek seems a lot less fine. 

It doesn't have to be all or nothing, either. One of the things we wish we had in the controller protocol is the ability to time out replicas that were down for too long, for some tweakable definition of "too long." Maybe something similar makes sense here, where peeks by default hold back the catalog revision, but the controller has license to abort them if they're taking too long, or as a matter of policy if certain commands come in (e.g., we decide that `ALTER TABLE` gets to force kill all outstanding peeks).

Reply:
> Oh this seems like a good thing to shake out!

agreed! this is subtle, though, so let's discuss in our next sync and then report back here?

Reply:
Sounds good!

Reply:
> But what if collection definitions can change over time?

We discussed this a bit in the scope of `ALTER MATERIALIZED VIEW` and arrived at the conclusion that having collection definitions immutable in the controller is very much preferable, as it simplifies things enormously. For example, if dataflow definitions can change in-place, you need to switch them out on the replicas. Do you wait for all replicas to have confirmed the switch? What if some replicas are unresponsive? Do you instead continue in a state where some replicas have switched the dataflow but some still send responses for the old one?

I don't think it would be too hard to keep collections immutable in the controller. If we can't re-associate the SQL object with a different `GlobalId`, we could do what @benesch suggested above and use a revision ID instead.